pkg/policy/resolve.go

```diff
@@ -157,40 +157,91 @@ func (p *EndpointPolicy) Detach() {
 
 // computeDesiredL4PolicyMapEntries transforms the EndpointPolicy.L4Policy into
 // the datapath-friendly format inside EndpointPolicy.PolicyMapState.
-func (p *EndpointPolicy) computeDesiredL4PolicyMapEntries() {
-
+// Called with selectorcache locked for reading
+func (p *EndpointPolicy) toMapState() {
 	if p.L4Policy == nil {
 		return
 	}
-	p.computeDirectionL4PolicyMapEntries(p.PolicyMapState, p.L4Policy.Ingress, trafficdirection.Ingress)
-	p.computeDirectionL4PolicyMapEntries(p.PolicyMapState, p.L4Policy.Egress, trafficdirection.Egress)
+	p.L4Policy.Ingress.toMapState(p)
+	p.L4Policy.Egress.toMapState(p)
 }
 
-func (p *EndpointPolicy) computeDirectionL4PolicyMapEntries(policyMapState MapState, l4PolicyMap L4PolicyMap, direction trafficdirection.TrafficDirection) {
-	for _, filter := range l4PolicyMap {
+// Called with selectorcache locked for reading
+func (l4policy L4DirectionPolicy) toMapState(p *EndpointPolicy) {
+	for _, l4 := range l4policy.PortRules {
 		lookupDone := false
 		proxyport := uint16(0)
-		keysFromFilter := filter.ToMapState(p.PolicyOwner, direction, p.SelectorCache)
-		for keyFromFilter, entry := range keysFromFilter {
+		l4.toMapState(p, l4policy.features, func(keyFromFilter Key, entry *MapStateEntry) bool {
 			// Fix up the proxy port for entries that need proxy redirection
 			if entry.IsRedirectEntry() {
 				if !lookupDone {
 					// only lookup once for each filter
 					// Use 'destPort' from the key as it is already resolved
 					// from a named port if needed.
-					proxyport = p.PolicyOwner.LookupRedirectPortLocked(filter.Ingress, string(filter.Protocol), keyFromFilter.DestPort)
+					proxyport = p.PolicyOwner.LookupRedirectPortLocked(l4.Ingress, string(l4.Protocol), keyFromFilter.DestPort)
 					lookupDone = true
 				}
 				entry.ProxyPort = proxyport
 				// If the currently allocated proxy port is 0, this is a new
 				// redirect, for which no port has been allocated yet. Ignore
 				// it for now. This will be configured by
-				// e.addNewRedirectsFromDesiredPolicy() once the port has been allocated.
+				// UpdateRedirects() once the port has been allocated.
 				if !entry.IsRedirectEntry() {
-					continue
+					return false
 				}
 			}
-			policyMapState.DenyPreferredInsert(keyFromFilter, entry, p.SelectorCache)
+			return true
+		}, nil, nil)
+	}
+}
+
+type getProxyPortFunc func(*L4Filter) (proxyPort uint16, ok bool)
+
+// UpdateRedirects updates redirects in the policy by calling the given 'filter' to obtain a proxy port number to use.
```

Comment:
nit, getting the proxy port isn't a filter by any meaning of `filter` I'm aware of: 
```suggestion
// UpdateRedirects updates redirects in the policy using the provided function to obtain a proxy port number to use.
```

Reply:
Also now that there's `adds` and `updated`, these should probably also be documented. It's becoming less clear to me whether this function modifies the received object or if it's supposed to be a function that doesn't change `p` but just uses info from `p` to populate `adds`/`updated`.

Reply:
Changed to:
```
// UpdateRedirects updates redirects in the EndpointPolicy's PolicyMapState by using the provided
// function to obtain a proxy port number to use. Changes to 'p.PolicyMapState' are collected in
// 'adds' and 'updated' so that they can be reverted when needed.
func (p *EndpointPolicy) UpdateRedirects(ingress bool, identities Identities, getProxyPort getProxyPortFunc, adds Keys, updated MapState) {
```